app/src/test/js/keysCtrlTest.js

```diff
@@ -0,0 +1,158 @@
+                expect(scope.crumbs[1]).toEqual(response.tree.children[0]);
+                expect(scope.parent).toEqual(response.tree.children[0]);
+
+                scope.manageChildren(scope.parent.children[0]);
+                expect(scope.crumbs.length).toEqual(3);
+                expect(scope.crumbs[0]).toEqual(response.tree);
+                expect(scope.crumbs[1]).toEqual(response.tree.children[0]);
+                expect(scope.crumbs[2]).toEqual(
+                        response.tree.children[0].children[0]);
+                expect(scope.parent).toEqual(
+                        response.tree.children[0].children[0]);
+
+                scope.gotoCrumb(scope.crumbs[0]);
+                expect(scope.crumbs.length).toEqual(1);
+                expect(scope.crumbs[0]).toEqual(response.tree);
+                expect(scope.parent).toEqual(response.tree);
+            }
+
+            it('crumb and manage chidren tests', function() {
+                crumbTest(syncKeysScope, syncKeysSuccessResponse);
+                crumbTest(searchKeysScope, searchKeysSuccessResponse);
+            });
+
+            function addTest(scope, response, name, type, value) {
+                expect(scope.addMode).toEqual(false);
+                expect(scope.crumbs.length).toEqual(1);
+                scope.addChild(scope.parent);
+                expect(scope.addMode).toEqual(true);
+                expect(scope.crumbs.length).toEqual(2);
+                scope.addFormName = name;
+                scope.addFormValue = value;
+                scope.requestAdd();
+                http.expectPOST(
+                        "api/key/addChild/" + scope.parent.id + "?name=" + name
+                                + "&type=" + type + "&value=" + value).respond(
+                        response);
+                http.flush();
+                expect(scope.parent).toEqual(response.result);
+                expect(scope.entities || scope.searchKeys).toEqual(
+                        response.tree);
+            }
+
+            it('add tests', function() {
+                addTest(syncKeysScope, syncKeysSuccessResponse, "testName",
+                        "Entity", "testValue");
+                addTest(searchKeysScope, searchKeysSuccessResponse,
+                        "SearchKeys", "SearchKey", "testValue");
+            });
+
+            function updateTests(scope, response, name, type, value) {
+                expect(scope.updateMode).toEqual(false);
+                expect(scope.crumbs.length).toEqual(1);
+                scope.editChild(scope.parent.children[0]);
+                expect(scope.updateMode).toEqual(true);
+                expect(scope.crumbs.length).toEqual(2);
+                expect(scope.parent).toEqual(response.tree.children[0]);
+                scope.parent.name = name;
+                scope.parent.value = value;
+                scope.requestUpdate();
+                http.expectPUT(
```

Comment:
I looked into stubbing out async calls some more and it turns out to be really easy, use `$q.when(value)` to return a promise that completes with `value` on the next digest cycle.

So if you wanted to stub out a service call...

```js
spyOn(myService, "getStuff").and.returnValue($q.when("test stuff"));
```

Or if you want to keep it simple you can just create a fake service object yourself:

```js
var myService = {
  getStuff: function() { return $q.when("test stuff"); }
};
```